Add support for dynamic tag names in html template literals

tbz opened the pull request:
## Description

Fixes #5922 by reusing the placeholder index on variables and member expressions that match each other.

## Checklist

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

<!-- Please DO NOT remove the playground link -->

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

fisker commented:
**Prettier pr-15545**
[Playground link](https://deploy-preview-15545--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEALGBbANgAwDpQAEhAPACTACGAFAJQB0AHgL4B8BxBJA9BQEZ0mbAjhAAaEBAAOMAJbQAzslCUATqogB3AApqESlJSybKATyUS+qymADWcGAGVKGOABlZUOMgBmRhXCW1nYOjlI2ngDmyDCqAK6BIAEYsjHxiQpRWHAAinEQ8L7+iQBWCoyOWbn5hUh+WAESAI41cNoaUgYglAoAtF5wACZD4iCxlLJYUQDCEBgYlMjdWFijmVCR2QCCMLGyfHHw2nCqHl5FDYno2ADqqLLwCuFgcI76D7IAbg+mS2AKFhAnwSAEkoMNYI4wKpZDItuDHDBTNkLo0QFINAEbtYpEsMXAAqpPt4JJ5CTB2pRIgtUYlwqpCUs+JQ+HBVhIMZ4YDdZIMYKhkAAOAAMElUcBasnFlOpizqxQkMBZPL5AqQACYJHEAgAVFkGepouAYVmDYaDNyUDZxKlwABiEFUC12USWlEOEBAzGYQA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
html`
  <${a().x}>
  
</${b().x}>
`
```

**Output:**
<!-- prettier-ignore -->
```jsx
html` <${a().x}> </${a().x}> `;

```

fisker reviewed (changes requested):
We need make sure the MemberExpression's object can be stringified.

tbz commented:
Thanks for the reviews! I tried to find a helper function for rendering the expression but couldn't find any. So I just added a check that the ME's object is one of the supported cases.

fisker commented:
Please add test for your change.

fisker commented:
New  bug

**Prettier pr-15545**
[Playground link](https://deploy-preview-15545--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEALGBbANgAwDpQAEhAPACTACGAFAJQB0Y9AHgL4B8BxBJA9BQCM6jFhwI4QAGhAQADjACW0AM7JQlAE4aIAdwAKmhKpSUsOygE9V0gRspgA1nBgBlShjgAZBVDjIAZqbKcDZ2js4usvY+AObIMBoAriEgwRgK8UkpyrFYcACKiRDwAUEpAFbKzC65BUUlSIFYwdIAjvVwetqyxiCUygC0vnAAJqNSIAmUClixAMIQGBiUyH1YWBM5UDF5AIIwCQoCifB6cBrevqXNKejYAOqoCvDKUWBwLkbPCgBuzxarMDKawgH7JACSUDGsBcYA0CnkuyhLhgFjy1xaIFk2mC9zsslW2LgwQ0Pz80h8JJgXUoMWWGJSUQ0JNWAkoAjgG2k2J8MHuChGMFQyAAHAAGaQaODtBRSml0laNMrSGDs-mC4VIABM0kSwQAKuzjE1MXAMByRmMRp5KNtErS4AAxCAaZYHWKrSgnCAgVisIA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
html`
  <${a().c.x}>
  
</${b().c.x}>
`
```

**Output:**
<!-- prettier-ignore -->
```jsx
html` <${a().c.x}> </${a().c.x}> `;

```

fisker commented on src/language-js/embed/html.js:
```diff
@@ -12,15 +12,50 @@ import {
+    return "this";
+  }
+
+  if (expression.type === "MemberExpression") {
```
Let's ignore computed and optional(just ti be safe)

**Prettier pr-15545**
[Playground link](https://deploy-preview-15545--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEALGBbANgAwDpQAEhAPACTACGAdAB4C+AfCQPQWUDatAukwflBAAaEBAAOMAJbQAzslCUAToogB3AApKEclJSyrKATzkiARospgA1nBgBlShjgAZSVDjIAZnplwzF61s7MUs3AHNkGEUAVz8QXwxJSJi4mXCsOABFaIh4Lx84gCsZWjt0rJy8pG8sXxEAR0q4dRUxHRBKGQBadzgAE37hECjKSSxwgGEIDAxKZA6sLCG0qDCMgEEYKMlTaPh1OEVXd3zauPRsAHVUSXgZELA4O21byQA3W8N5sBkTEDfYgBJKADWB2MCKSQSdYguwwQwZU51EBiFS+S4WMTzVFwXyKN4eERuPEwFqUMKzJFxEKKPHzUyUUxwJYiVFuGCXSR9GCoZAADgADCJFHBGpIRWSKXNqgURDBGZzubykAAmETRXwAFUZOhqyLgGCZfQGfWclFW0XJcAAYhBFLMtuF5pQ9hAQPR6EA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
html`
  <${a.x}></${a[x]}>
`

```

**Output:**
<!-- prettier-ignore -->
```jsx
html` <${a.x}></${a.x}> `;

```

fisker commented:
We don't want duplicate or loose comments either

**Prettier pr-15545**
[Playground link](https://deploy-preview-15545--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEALGBbANgAwDpQAEhAPACTAwCGA5oQPQBUhkGGCMhAjIY-QL4A+EvQrUaQgvigF02acXKVaQkRSYsIbDoQBMveuMlRpIADQgIABxgBLaAGdkoKgCdXEAO4AFNwicoVFieVACeThYARq5UYADWcDAAylTsADK2UHDIAGZBDnBRMfGJSVaxmTTIMK4AroUgBRi21XUNDpVYcACKtRDwufkNAFYOAB5JnT19A0h5WAUWAI4zcN4eVgEgVA4AtFlwACZH5iA1VLZYlQDCWhhUyNtYWKcdUDRdAIIwNbaRtfBvHBXBksoMFg05FgAOqoWzwBzlMBwJL+eG2ABu8NCjzADgiIAx9QAklBjrAkmBXLYbJ8yUkYKEuuDFiArB4CtCYlZHuy4AVXBjshZMgKYOtaPcWQ1yq4BY9IlRInAXhZ2ZkYNDbIcYKhkAAOAAMFlccBWtlNEpoUrmQws1EiWp1eqQugstQKABUlQF5qy4BhlYdjoc0lR3rVaHAAGIQVz3H6VR5UAEQED8fhAA)
<!-- prettier-ignore -->
```sh
--parser babel
```

**Input:**
<!-- prettier-ignore -->
```jsx
html`
  <${tag /* comment 1 */}></${tag}>
`

html`
  <${tag}></${/* comment 2 */tag}>
`

```

**Output:**
<!-- prettier-ignore -->
```jsx
html` <${tag /* comment 1 */}></${tag /* comment 1 */}> `;

html` <${tag}></${tag}> `;

```

tbz commented:
Sorry for the delay, I have updated the PR so that we ignore comments, and computed/optional members. Also added support for Lit's unsafeStatic() and other method calls.
